src/data/polynomial/lifts.lean

```diff
@@ -0,0 +1,332 @@
+lemma X_pow_lifts (f : R →+* S) (n : ℕ) : lifts f (X ^ n : polynomial S) :=
+  by use X ^ n; rw [map_pow, map_X]
+
+/--If `p` and `q` lift then `p + q` lifts. -/
+lemma lifts_add_mem {p q : polynomial S} (hp : lifts f p) (hq : lifts f q) : lifts f (p + q) :=
+  by rw lifts_iff at hp hq ⊢; exact subsemiring.add_mem (ring_hom.of (map f)).srange hp hq
+
+/--If `p` and `q` lift then `p * q` lifts. -/
+lemma lifts_mul_mem {p q : polynomial S} (hp : lifts f p) (hq : lifts f q) : lifts f (p * q) :=
+  by rw lifts_iff at hp hq ⊢; exact subsemiring.mul_mem (ring_hom.of (map f)).srange hp hq
+
+/--If `p` lifts and `(n : ℕ)` then `p ^ n` lifts. -/
+lemma lifts.pow_mem {p : polynomial S} (n : ℕ) (hp : lifts f p) : lifts f (p ^ n) :=
+  by rw lifts_iff at hp ⊢; exact subsemiring.pow_mem (ring_hom.of (map f)).srange hp n
+
+/--If `p` lifts and `(n : ℕ)` then `n • p` lifts. -/
+lemma lifts.nat_smul_mem {p : polynomial S} (n : ℕ) (hp : lifts f p) : lifts f (n • p) :=
+  by rw lifts_iff at hp ⊢; exact subsemiring.nsmul_mem (ring_hom.of (map f)).srange hp n
+
+/--If `p` lifts and `(r : R)` then `r * p` lifts. -/
+lemma lifts.base_mul_mem {p : polynomial S} (r : R) (hp : lifts f p) : lifts f (C (f r) * p) :=
+  by rw lifts at hp ⊢; obtain ⟨p₁, rfl⟩ := hp; use C r * p₁; rw [map_mul, map_C]
+
+/--If any element of a list of polynomials lifts, then the product lifts-/
+lemma lifts_list_prod_mem {L : list (polynomial S)} :
+  (∀ (p : polynomial S), p ∈ L → lifts f p) → lifts f L.prod :=
+  by simp_rw [lifts_iff]; exact subsemiring.list_prod_mem (ring_hom.of (map f)).srange
+
+/--If any element of a list of polynomials lifts, then the sum lifts-/
+lemma lifts_list_sum_mem {L : list (polynomial S)} :
+  (∀ (p : polynomial S), p ∈ L → lifts f p) → lifts f L.sum :=
+  by simp_rw [lifts_iff]; exact subsemiring.list_sum_mem (ring_hom.of (map f)).srange
+
+/--If any element of a multiset of polynomials lifts, then the sum lifts-/
+lemma lifts_multiset_sum_mem {m : multiset (polynomial S)} :
+  (∀ (p : polynomial S), p ∈ m → lifts f p) → lifts f m.sum :=
+  by simp_rw [lifts_iff]; exact subsemiring.multiset_sum_mem (ring_hom.of (map f)).srange m
+
+/--If any element of a finset of polynomials lifts, then the sum lifts-/
+lemma lifts_sum_mem {ι : Type w} {t : finset ι} {g : ι → (polynomial S)} :
+  (∀ (x : ι), x ∈ t → lifts f (g x)) → lifts f (∑ (x : ι) in t, g x) :=
+by simp_rw [lifts_iff]; exact subsemiring.sum_mem (ring_hom.of (map f)).srange
+
+/--If `p` lifts then is leading monomial lifts. -/
+lemma lead_monom_lifts_of_lifts {p : polynomial S} (h : lifts f p) :
+  lifts f ((C p.leading_coeff) * polynomial.X ^ p.nat_degree) :=
+begin
+  obtain ⟨q, hq⟩ := h,
+  rw leading_coeff,
+  use (C (q.coeff p.nat_degree)) * polynomial.X ^ p.nat_degree,
+  nth_rewrite 2 ←hq,
+  simp only [map_C, coeff_map, map_pow, map_X, map_mul]
+end
+
+/--If `p` lifts then `p.erase_lead` lifts. -/
+lemma erase_lead_lifts_of_lifts {p : polynomial S} (h : lifts f p) : lifts f p.erase_lead :=
```

Comment:
We have a good test for `leads` near the definition of `map`: every coefficient belongs to the range of `f`. This immediately implies that every `p.erase k` lifts.